### PR TITLE
Issue/30

### PR DESCRIPTION
**Pull request by ghost**
- Added a dialog menu for changing interface setting.
- Added Error popups for the interface settings menu.
- It is now possible to change the Flywheel interface device on the fly.
- Changes to the device are recorded in the log output.
- Flywheel and graphs now run at different refresh rates.
- Refresh rate code appears to work correctly
*note: Timers round to the nearest 1ms which cause the 31 fps error in the fps output.

### CODE REVIEW

File: FESS-GUI/setpassworddialog.cpp [hunk 1]
```diff
@@ -13,6 +14,7 @@ SetPasswordDialog::SetPasswordDialog(QWidget *parent) :
     ui(new Ui::SetPasswordDialog)
 {
     ui->setupUi(this);
+    qsrand(time(NULL));
```
Comment (JCJutson) [inline FESS-GUI/setpassworddialog.cpp]: This causes a compile error on my machine because time is not declared. Include ctime if you want to do this.

Comment (Zyrck56) [reply FESS-GUI/setpassworddialog.cpp]: Fixed in my branch.

Comment (JCJutson) [reply FESS-GUI/setpassworddialog.cpp]: Nice. On a broader note, I don't agree with moving this from the mainwindow constructor to the setpassworddialog constructor. qrand() is used in GetRandomString, which is not a member of setpassworddialog. So it isn't guaranteed that a setpassworddialog will be constructed before that funtion is used (if it is called from outside the class). Furthermore, qsrand only needs to be set up once; setting it up every time a setpassworddialog is created is unneccessary. If anything, qsrand(time(NULL)) should be moved to main(), the entry point of the program.

File: FESS-GUI/FESS-GUI.pro [hunk 2]
```diff
@@ -18,32 +18,37 @@ SOURCES     +=  main.cpp \
...
                 recordingoperation.cpp \
-                interfacemanager.cpp \
                 transmitbuffer.cpp \
                 conversions.cpp \
                 flypacket.cpp \
                 demodevice.cpp \
                 serialdevice.cpp \
-                graph.cpp
+                graph.cpp \
+    commoninterfaceselector.cpp \
```
Comment (JCJutson) [inline FESS-GUI/FESS-GUI.pro]: Let's try to keep these at the same indentation level.

File: FESS-GUI/mainwindow.cpp [hunk 3]
```diff
@@ -1,39 +1,47 @@
...
+    setTimers();
+    setUpSignals();
+    setUpKeyBindings();
+
+    this->setWindowTitle(MAINWINDOW_TITLE);
+
+    graphRefreshRate = 100;
+    flywheelRefreshRate = 200;
+
+    graphRefreshTimer->setInterval(refreshRateToMS(graphRefreshRate));
```
Comment (JCJutson) [inline FESS-GUI/mainwindow.cpp]: I think it would be best to move this line, line 36, and line 39 into SetTimers(). If we're going to have a SetTimers() function, may as well move all timer setup code into it.

File: FESS-GUI/mainwindow.cpp [hunk 4]
```diff
@@ -86,9 +91,7 @@ MainWindow::MainWindow(QWidget *parent) :
     rotationGraph = new LocationGraph(ui->mainRotGraph, ui->auxRotatGraph, std::vector<QColor>{rotationalColor}, "mm", 1);
 
     // setup a timer that repeatedly calls MainWindow::realtimeDataSlot:
```
Comment (JCJutson) [inline FESS-GUI/mainwindow.cpp]: Remove these comments, since they no longer apply here.

File: FESS-GUI/mainwindow.cpp [hunk 5]
```diff
@@ -100,8 +103,7 @@ MainWindow::MainWindow(QWidget *parent) :
 
 void MainWindow::realtimeDataSlot()  //Important function. This is repeatedly called
 {
-    flywheelOperation->sync();
-    //at the refresh rate defined by the timer (line 107)
+    //at the refresh rate defined by the timer
```
Comment (JCJutson) [inline FESS-GUI/mainwindow.cpp:141]: sync is no longer needed?

File: FESS-GUI/mainwindow.cpp [hunk 6]
```diff
@@ -1,39 +1,47 @@
...
     ui->setupUi(this);
 
-    qsrand(time(NULL));
+    setTimers();
+    setUpSignals();
+    setUpKeyBindings();
+
+    this->setWindowTitle(MAINWINDOW_TITLE);
+
+    graphRefreshRate = 100;
```
Comment (JCJutson) [inline FESS-GUI/mainwindow.cpp]: Let's be consistent with where we initialize variables to constants. Before, we were doing this in the header file.

File: FESS-GUI/mainwindow.cpp [hunk 7]
```diff
@@ -522,19 +518,157 @@ void MainWindow::on_lineEditPassword_textEdited(const QString &password)
...
-
-
 void MainWindow::on_actionLock_frame_rate_at_30FPS_triggered(bool checked)
 {
-    dataTimer->stop();
     if(checked)
-        refreshRate = 33;
+    {
+        graphRefreshRate = 30;
+        graphRefreshTimer->setInterval(refreshRateToMS(graphRefreshRate));
```
Comment (JCJutson) [inline FESS-GUI/mainwindow.cpp]: Why duplicate this line in both blocks? Having it after the end of the if and else blocks cuts down on duplicated code.

File: FESS-GUI/mainwindow.cpp [hunk 8]
```diff
@@ -522,19 +518,157 @@ void MainWindow::on_lineEditPassword_textEdited(const QString &password)
...
+void MainWindow::runFlywheelOperations()
+{
+    flywheelOperation->sync();
+}
+
+void MainWindow::startFlywheelInterface()
+{
+    deviceInterface = interfaceManager->getCurrentInterface();
+
+    if(!deviceInterface)
```
Comment (JCJutson) [inline FESS-GUI/mainwindow.cpp]: Is this equivalent to deviceInterface == NULL ? If so, let's change it to that. This is a little confusing for me

File: FESS-GUI/setpassworddialog.cpp (hunk 1 above)
Comment (ghost) [reply FESS-GUI/setpassworddialog.cpp]: No problem, we can move it back. I was cleaning up libraries and it compiled and ran fine on my computer so I had not clue ctime was needed.

File: FESS-GUI/commoninterfacemanager.cpp [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+CommonDeviceInterface* CommonInterfaceManager::getCurrentInterface()
+{
+    return currentDevice;
+}
+
+ bool CommonInterfaceManager::isADeviceSet()
+ {
+     if(currentDevice)
```
Comment (JCJutson) [inline FESS-GUI/commoninterfacemanager.cpp]: how about "return currentDevice != NULL"?

File: FESS-GUI/commoninterfaceselector.h [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    void demoButtonSetClicked();
+    void demoButtonCancelClicked();
+
+    void serialButtonSetClicked();
+    void serialButtonCancelClicked();
+
+private:
+    Ui::CommonInterfaceSelector *ui;
+
+    bool serial_port_text_modified;
```
Comment (JCJutson) [inline FESS-GUI/commoninterfaceselector.h]: underscores instead of camelCase? :(

File: FESS-GUI/demodevice.cpp [hunk 11]
```diff
@@ -144,16 +148,27 @@ void DemoDevice::sync()
 // Interface Overedload Functions
 //--------------------------------------------------------------------
 
+bool DemoDevice::ready()
```
Comment (JCJutson) [inline FESS-GUI/demodevice.cpp]: the convention for boolean names is is_______. So here, isReady(). It just makes the code read a bit more like english. e.g. if(device->isReady()){...}

File: FESS-GUI/flywheeloperation.cpp [hunk 12]
```diff
@@ -177,90 +178,93 @@ void FlywheelOperation::setDefaults()
...
             }
-        }
 
-       // if (( emergency_retries > 0 )&&( emergency_acknowlegded == false ))
-       // {
-       //     communication_device->pushCommand(COMMAND_SET_EMER_STOP);
-       //     communication_device->sync();
-       //     emergency_retries--;
-       // }
+           // if (( emergency_retries > 0 )&&( emergency_acknowlegded == false ))
```
Comment (JCJutson) [inline FESS-GUI/flywheeloperation.cpp]: Is there a todo here? It's generally not a good idea to merge commented code to master

File: FESS-GUI/graph.cpp [hunk 13]
```diff
@@ -154,7 +154,7 @@ void LocationGraph::setupPlot(QCustomPlot *plot, bool isMain, std::vector<QColor
 
 void LocationGraph::addData(std::vector<QPointF> points)  //function to add xy data to graphs
 {
-    for(int i = 0; i < points.size() && i < numPoints; i++){
+    for(unsigned int i = 0; i < points.size() && i < (unsigned int)numPoints; i++){
```
Comment (JCJutson) [inline FESS-GUI/graph.cpp]: Why this change? Currently, there are either numPoints is either 1 or 2.

File: FESS-GUI/mainwindow.cpp [hunk 14]
```diff
@@ -231,7 +237,7 @@ void MainWindow::on_goButton_clicked()  //when you hit the go button
         goplayer->play();
     }
     //Pass information on to text browser to be displayed
-    ui->textBrowser->append(QString("Flywheel controlled to %1 RPM,"
+    ui->outputLog->append(QString("Flywheel controlled to %1 RPM,"
```
Comment (JCJutson) [inline FESS-GUI/mainwindow.cpp:271]: Solid renaming. 👍

File: FESS-GUI/mainwindow.cpp [hunk 15]
```diff
@@ -522,19 +522,160 @@ void MainWindow::on_lineEditPassword_textEdited(const QString &password)
...
+
+void MainWindow::on_actionSet_Reset_Password_triggered() //show the password dialog box
+{
+    SetPasswordDialog* passwordResetWindow = new SetPasswordDialog(this);
+    passwordResetWindow->show();
+}
+
+// Signal Setups
+
+void MainWindow::setUpSignals()
```
Comment (JCJutson) [inline FESS-GUI/mainwindow.cpp]: Move these to right after the constructor, so all constructor code is in one place

File: FESS-GUI/windowsnames.h [hunk 16]
```diff
@@ -0,0 +1,7 @@
+#ifndef WINDOWSNAMES
+#define WINDOWSNAMES
+
+#define MAINWINDOW_TITLE "FlyCAM - Flywheel Control and Monitoring"
```
Comment (JCJutson) [inline FESS-GUI/windowsnames.h:4]: I usually shy away from #defines in C++. string MainWindowTitle = "..." would be a preferable way to do this. I think that's the more common way to do it in C++, as opposed to in C, where #defines are a bit more common.

**Review by JCJutson — CHANGES_REQUESTED** [review]
Great feature, Ben. A lot of good work is represented in this code. I have a few small changes to request, just in naming and the like to keep with the conventions we've followed so far in the code. This will help keep our codebase consistent and easy for a future team to pick up and understand.

File: FESS-GUI/mainwindow.cpp (hunk 8 above)
Comment (Zyrck56) [reply FESS-GUI/mainwindow.cpp]: Have to disagree on this one 

"In C and C++ programming, two null pointers are guaranteed to compare equal; ANSI C guarantees that any null pointer will be equal to 0 in a comparison with an integer type; furthermore the macro NULL is defined as a null pointer constant, that is value 0 (either as an integer type or converted to a pointer to void), so a null pointer will compare equal to NULL."

Comment (JCJutson) [reply FESS-GUI/mainwindow.cpp]: You think (!deviceInterface) is more readable than (deviceInterface == NULL) ?

Comment (Zyrck56) [reply FESS-GUI/mainwindow.cpp]: No. Just saying its a valid comparison.

Comment (JCJutson) [reply FESS-GUI/mainwindow.cpp]: I'm sure it is. My comment was only about readability.

Comment (Zyrck56) [reply FESS-GUI/mainwindow.cpp]: Alright didn't realize

File: FESS-GUI/mainwindow.cpp (hunk 5 above)
Comment (ghost) [reply FESS-GUI/mainwindow.cpp:141]: Sync is now bound to a signal from a timer so that it can run independently of the graphs.

File: FESS-GUI/mainwindow.cpp (hunk 4 above)
Comment (ghost) [reply FESS-GUI/mainwindow.cpp]: Done

File: FESS-GUI/mainwindow.cpp (hunk 3 above)
Comment (ghost) [reply FESS-GUI/mainwindow.cpp]: Good call. Will do.

File: FESS-GUI/mainwindow.cpp (hunk 7 above)
Comment (ghost) [reply FESS-GUI/mainwindow.cpp]: No reason really, just a simple copy, paste and modify when I wrote it. Moved it outside the else to decrease duplication.

File: FESS-GUI/commoninterfaceselector.h (hunk 10 above)
Comment (ghost) [reply FESS-GUI/commoninterfaceselector.h]: I typically use lower case and underscores for built-in data types and camelCase for objects but I can change it to camelCase to be consistent with the rest of the project.

File: FESS-GUI/flywheeloperation.cpp (hunk 12 above)
Comment (ghost) [reply FESS-GUI/flywheeloperation.cpp]: Oops, that is the emergency stop to-do code. I was suppose to be in another branch.

File: FESS-GUI/graph.cpp (hunk 13 above)
Comment (ghost) [reply FESS-GUI/graph.cpp]: Tom needs to answer that one but I think the type casting was done to fix a compiler warning.

Comment (Zyrck56) [reply FESS-GUI/graph.cpp]: size() returns an unsigned int so I was getting this annoying comparison warning when I compiled the program.. I guess if no one else has that issue we can revert this.

Comment (Zyrck56) [reply FESS-GUI/graph.cpp]: Comes from the fact that an unsigned int can be out of the bounds of a signed int. What c++ compiler are you guys using?

File: FESS-GUI/windowsnames.h (hunk 16 above)
Comment (ghost) [reply FESS-GUI/windowsnames.h:4]: It depends defines are typeless and don't use any memory until they are used, converted and processed by the compiler. They also don't need typecasting. If we defined all our text in string they would need to be cast to QString for use in QT. In both instances the program would need to be recompiled if the text in those files was changed. However if this text was stored in a localization file using variables would be a better option.

File: FESS-GUI/mainwindow.cpp (hunk 3 above)
Comment (ghost) [reply FESS-GUI/mainwindow.cpp]: Done

File: FESS-GUI/graph.cpp (hunk 13 above)
Comment (ghost) [reply FESS-GUI/graph.cpp]: I am using g++ which has the warning if i is just an int.

**Review by JCJutson — APPROVED** [review]

**Comment by JCJutson** [comment]
Looks great. Good work you guys
